pkg/cmd/monitor/health/health.go

```diff
@@ -0,0 +1,156 @@
+// HashKVRequest and MemberList (3.4 and earlier).
+func WithSerializedReadSingleTarget() CheckFunc {
+	return func(ctx context.Context, c *Check) {
+		c.name = SerializedReadSingleTarget
+		c.checkKVSRead(ctx, []clientv3.OpOption{clientv3.WithSerializable(), clientv3.WithKeysOnly()})
+	}
+}
+
+// WithQuorumReadSingleTarget performs a quorate RangeRequest against a single target. This check is useful to
+// understand exact timings on quorate availability on a per peer level. This single target test can fail on a
+// minority of peer targets while WithQuorumRead passes. This can happen most commonly on node reboot during
+// upgrades or during a static-pod rollout. But can also happen less during a network partition.
+func WithQuorumReadSingleTarget() CheckFunc {
+	return func(ctx context.Context, c *Check) {
+		c.name = QuorumReadSingleTarget
+		c.checkKVSRead(ctx, []clientv3.OpOption{clientv3.WithKeysOnly()})
+	}
+}
+
+// WithGRPCReadySingleTarget performs a gRPC dial to the endpoint and checks if the service is "Ready". This check
+// ensures that the target subconn will be considered usable by the etcd client balancer for round robin requests. Ready
+// does not preclude quorum.
+func WithGRPCReadySingleTarget() CheckFunc {
+	return func(ctx context.Context, c *Check) {
+		c.name = GRPCReadySingleTarget
+		c.checkDialStatus(ctx)
+	}
+}
+
+// WithQuorumRead performs a quorate RangeRequest against all targets. This check ensures given an etcd client
+// with multiple endpoint targets the client balancer can provide linearized read/write service to the apiserver.
+func WithQuorumRead() CheckFunc {
+	return func(ctx context.Context, c *Check) {
+		c.name = QuorumRead
+		c.checkKVSRead(ctx, []clientv3.OpOption{clientv3.WithKeysOnly()})
+	}
+}
+
+func (c *Check) checkDialStatus(_ context.Context) {
+	c.ensureCheckStatus()
+	// Due to client retry failure will not be observed directly.
+	slowRequestTimer := time.AfterFunc(DefaultSlowRequestDuration, func() {
+		c.logSlowRequest(fmt.Errorf("slow request"))
+	})
+	defer slowRequestTimer.Stop()
+
+	resp, err := c.client.Dial(c.targets[0])
+	if err != nil {
+		c.logHealthFail(err)
+		return
+	}
+	defer resp.Close()
+	if resp.GetState() == connectivity.Ready {
+		c.logHealthSuccess()
+		return
+	}
+	c.logHealthFail(fmt.Errorf("grpc: ClientConn state %q", resp.GetState()))
+}
+
+func (c *Check) checkKVSRead(ctx context.Context, opts []clientv3.OpOption) {
+	c.ensureCheckStatus()
+
+	//Due to client retry failure will not be observed directly.
+	slowRequestTimer := time.AfterFunc(DefaultSlowRequestDuration, func() {
+		c.logSlowRequest(fmt.Errorf("slow request"))
+	})
+	defer slowRequestTimer.Stop()
+
+	// Without a context timeout we will retry non mutating request indefinitely.
+	ctx, cancel := context.WithTimeout(ctx, DefaultContextDeadline)
+	defer cancel()
+	// pin targets for single target tests
+	if len(c.targets) == 1 {
+		c.client.SetEndpoints(c.targets...)
```

Comment:
This call here. You're sharing the one client you create here.  If you dropped this line, I wouldn't worry about.  Do you need this line?

Reply:
I create one client per target so each slice of checks has the same endpoints. While we set we set the same value. I create each client with 3 endpoints. The reason being we will fail to startup if that single endpoint doesn't respond. So in the case of upgrade we could be in fail loop. 

```
for each []targets {
   create client
   for each []check {
     consume client
```

Reply:
I can start and set early will fix it.